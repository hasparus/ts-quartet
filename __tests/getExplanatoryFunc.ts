export const getExplanatoryFunc = (
  validValue: string,
  explanation: string
) => () => ({
  check: (v: any) => `${v} === ${JSON.stringify(validValue)}`,
  not: (v: any) => `${v} !== ${JSON.stringify(validValue)}`,
  handleError: (v: any, ctx: any) =>
    `${ctx}.explanations.push(${JSON.stringify(explanation)})`
});
