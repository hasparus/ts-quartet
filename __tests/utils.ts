import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
