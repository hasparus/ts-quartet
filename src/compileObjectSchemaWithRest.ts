import { addTabs } from "./addTabs";
import { compileIfNotValidReturnFalse } from "./compileIfNotValidReturnFalse";
import { CompilationResult, IObjectSchema, Prepare, QuartetInstance } from "./types";

export function compileObjectSchemaWithRest(
  v: QuartetInstance,
  s: IObjectSchema
): CompilationResult {
  const preparations: Prepare[] = [];
  const [funcBodyCode, isPure] = compileIfNotValidReturnFalse(
    v,
    "value",
    "validator",
    s,
    preparations
  );
  // tslint:disable-next-line
  const ctx = eval(
    `
      (()=>{
        function validator(value) {${
          isPure ? "" : "\n  validator.explanations = []"
        }\n${addTabs(funcBodyCode)}\n  return true\n}
        return validator
      })()
    `
  );
  ctx.explanations = [];
  ctx.pure = isPure;
  // tslint:disable-next-line
  for (let i = 0; i < preparations.length; i++) {
    preparations[i](ctx);
  }
  return ctx;
}
