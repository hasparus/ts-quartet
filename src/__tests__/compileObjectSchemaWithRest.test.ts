import { compileObjectSchemaWithRest } from "../compileObjectSchemaWithRest";
import { v } from "../index";
import { getExplanatoryFunc } from "./getExplanatoryFunc";
import { snapshot, tables, getExplanation } from "./utils";

describe("compileObjectSchemaWithRest", () => {
  test("empty obj", () => {
    const validator = compileObjectSchemaWithRest(v, {
      [v.rest]: getExplanatoryFunc("A", "Is not A")
    });
    expect(getExplanation(validator, {})).toMatchInlineSnapshot(`Array []`);
    expect(getExplanation(validator, { a: "A" })).toMatchInlineSnapshot(
      `Array []`
    );
    expect(getExplanation(validator, { a: "A", b: "b" }))
      .toMatchInlineSnapshot(`
                              Array [
                                "Is not A",
                              ]
                    `);
    snapshot(validator);
  });
  test("obj with func", () => {
    const validator = compileObjectSchemaWithRest(v, {
      id: () => ({
        check: id => `${id} === 42`,
        not: id => `${id} !== 42`,
        handleError: (id, ctx) => `${ctx}.explanations.push(${id})`
      }),
      [v.rest]: getExplanatoryFunc("A", "Is not A")
    });
    tables(
      validator,
      [{ id: 42 }, { id: 42, other: "A" }],
      [
        {},
        { id: 42, other: "B" },
        { id: "42" },
        null,
        false,
        true,
        { id: false },
        "41 + 1"
      ]
    );
    expect(getExplanation(validator, { id: 43 })).toEqual([43]);
    expect(getExplanation(validator, { id: 42, other: "b" })).toEqual([
      "Is not A"
    ]);
    snapshot(validator);
  });
  test("obj with constant", () => {
    const validator = compileObjectSchemaWithRest(v, {
      id: 42,
      nan: NaN,
      inf: Infinity,
      minf: -Infinity,
      zero: -0,
      str: "Andrew",
      f: false,
      t: true,
      symb: Symbol.for("test"),
      [v.rest]: v.string
    });
    tables(
      validator,
      [
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        }
      ],
      [
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test"),
          other: 10
        },
        null,
        {},
        {
          id: "42",
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: "NaN",
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: "Infinity",
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: "-Infinity",
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: "0",
          str: "Andrew",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "",
          f: false,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: null,
          t: true,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: 1,
          symb: Symbol.for("test")
        },
        {
          id: 42,
          nan: NaN,
          inf: Infinity,
          minf: -Infinity,
          zero: 0,
          str: "Andrew",
          f: false,
          t: true,
          symb: "test"
        }
      ]
    );

    snapshot(validator);
  });
  test("obj with variants", () => {
    const validator = compileObjectSchemaWithRest(v, {
      gender: ["male", "female"],
      grade: Array.from({ length: 11 }, (_, i) => i + 1),
      [v.rest]: [
        getExplanatoryFunc("A", "Is not A"),
        getExplanatoryFunc("B", "Is not B")
      ]
    });

    const invalids = [
      {},
      { gender: "male" },
      { grade: 10 },
      { grade: "10" },
      { gender: "male", grade: 0 },
      { grade: 12, gender: "female" },
      { grade: "10" },
      { grade: "10", gender: "male" },
      { grade: 10, gender: "male", other: "a" }
    ];
    const valids = ([] as any).concat(
      ...["male", "female"].map(gender =>
        Array.from({ length: 11 }, (_, grade) => ({
          grade: grade + 1,
          gender,
          test: "A",
          other: "B"
        }))
      )
    );
    tables(validator, valids, invalids);
    expect(getExplanation(validator, { grade: 10, gender: "male", other: "a" }))
      .toMatchInlineSnapshot(`
                        Array [
                          "Is not A",
                          "Is not B",
                        ]
                `);
    snapshot(validator);
  });
  test("obj + obj", () => {
    const validator = compileObjectSchemaWithRest(v, {
      deep: {
        deep: {
          space: () => ({
            check: v => `${v} === 'true'`,
            handleError: (v, ctx) => `${ctx}.explanations.push(${v})`
          }),
          [v.rest]: getExplanatoryFunc("A", "Is not A")
        },
        [v.rest]: getExplanatoryFunc("B", "Is not B")
      }
    });
    snapshot(validator);
    tables(
      validator,
      [
        { deep: { deep: { space: "true", other: "A" } } },
        { deep: { deep: { space: "true" } } },
        { deep: { deep: { space: "true", other: "A" }, yetAnother: "B" } }
      ],
      [
        { deep: { deep: { space: "false" } } },
        { deep: { deep: { space: "true", other: "b" } } },
        { deep: { deep: { space: "true", other: "A" }, yetAnother: "c" } },
        { deep: { deep: { space: "true", other: "a" }, yetAnother: "B" } }
      ]
    );
    expect(getExplanation(validator, { deep: { deep: { space: "false" } } }))
      .toMatchInlineSnapshot(`
                                                            Array [
                                                              "false",
                                                            ]
                                        `);
    expect(
      getExplanation(validator, {
        deep: { deep: { space: "true" }, other: "b" }
      })
    ).toMatchInlineSnapshot(`
                  Array [
                    "Is not B",
                  ]
            `);
    expect(
      getExplanation(validator, {
        deep: { deep: { space: "true", other: "a" } }
      })
    ).toMatchInlineSnapshot(`
                  Array [
                    "Is not A",
                  ]
            `);
  });
  test("obj: variant explanations", () => {
    const validator = compileObjectSchemaWithRest(v, {
      var: [
        getExplanatoryFunc("A", "Is not A"),
        getExplanatoryFunc("B", "Is not B")
      ],
      [v.rest]: getExplanatoryFunc("C", "Is not C")
    });
    tables(
      validator,
      [{ var: "A" }, { var: "B" }, { var: "A", b: "C" }],
      [{ var: "C" }, { var: "A", B: "d" }, { var: "B", b: "d" }]
    );
    expect(getExplanation(validator, { var: "a" })).toMatchInlineSnapshot(`
            Array [
              "Is not A",
              "Is not B",
            ]
        `);
    expect(getExplanation(validator, { var: "A", b: "d" }))
      .toMatchInlineSnapshot(`
            Array [
              "Is not C",
            ]
        `);
    expect(getExplanation(validator, { var: "B", b: "d" }))
      .toMatchInlineSnapshot(`
            Array [
              "Is not C",
            ]
        `);
    snapshot(validator);
  });
  test("obj: variants explanations", () => {
    const validator = v([
      null,
      {
        a: getExplanatoryFunc("A", "Is not A"),
        [v.rest]: getExplanatoryFunc("C", "Is not C")
      }
    ]);

    tables(
      validator,
      [null, { a: "A" }, { a: "A", b: "C" }],
      [
        false,
        undefined,
        0,
        { a: "A", b: "c" },
        { a: "b" },
        { a: "a" },
        { a: "a" }
      ]
    );
    expect(getExplanation(validator, { a: "b" })).toMatchInlineSnapshot(`
                                          Array [
                                            "Is not A",
                                          ]
                            `);
    expect(getExplanation(validator, { a: "A" })).toMatchInlineSnapshot(
      `Array []`
    );
    expect(getExplanation(validator, { a: "A", b: "C" })).toMatchInlineSnapshot(
      `Array []`
    );
    expect(getExplanation(validator, { a: "A", b: "c" }))
      .toMatchInlineSnapshot(`
      Array [
        "Is not C",
      ]
    `);
  });
});
