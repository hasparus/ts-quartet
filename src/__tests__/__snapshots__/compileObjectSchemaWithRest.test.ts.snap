// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileObjectSchemaWithRest empty obj 4`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (!validator['rest-validator-0'](value[key])) {
        validator.explanations.push(...validator['rest-validator-0'].explanations)
        return false
      }
    }
    return true
  }",
  "defined-0": "function (value) { return value; }",
  "defined-0.explanations": Array [],
  "explanations": Array [
    "Is not A",
  ],
  "rest-validator-0": "function validator(value) {
  validator.explanations = []
  if (value === \\"A\\") {
    return true
  }
  validator.explanations.push(\\"Is not A\\")
  return false
}",
  "rest-validator-0.explanations": Array [
    "Is not A",
  ],
}
`;

exports[`compileObjectSchemaWithRest obj + obj 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-6'](value)) {
      validator.explanations.push(...validator['defined-6'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-4'](value[key])) {
        validator.explanations.push(...validator['rest-validator-4'].explanations)
        return false
      }
    }
    return true
  }",
  "__propsWithSchemasDict": Object {
    "deep": true,
  },
  "defined-6": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['deep-1'](value.deep)) {
      validator.explanations.push(...validator['deep-1'].explanations)
      return false
    }
    return true
  }",
  "defined-6.deep-1": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-5'](value)) {
      validator.explanations.push(...validator['defined-5'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-5'](value[key])) {
        validator.explanations.push(...validator['rest-validator-5'].explanations)
        return false
      }
    }
    return true
  }",
  "defined-6.deep-1.__propsWithSchemasDict": Object {
    "deep": true,
  },
  "defined-6.deep-1.defined-5": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['deep-0'](value.deep)) {
      validator.explanations.push(...validator['deep-0'].explanations)
      return false
    }
    return true
  }",
  "defined-6.deep-1.defined-5.deep-0": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-4'](value)) {
      validator.explanations.push(...validator['defined-4'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-6'](value[key])) {
        validator.explanations.push(...validator['rest-validator-6'].explanations)
        return false
      }
    }
    return true
  }",
  "defined-6.deep-1.defined-5.deep-0.__propsWithSchemasDict": Object {
    "space": true,
  },
  "defined-6.deep-1.defined-5.deep-0.defined-4": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!(value.space === 'true')) {
      validator.explanations.push(value.space)
      return false
    }
    return true
  }",
  "defined-6.deep-1.defined-5.deep-0.defined-4.explanations": Array [],
  "defined-6.deep-1.defined-5.deep-0.explanations": Array [],
  "defined-6.deep-1.defined-5.deep-0.rest-validator-6": "function validator(value) {
  validator.explanations = []
  if (value === \\"A\\") {
    return true
  }
  validator.explanations.push(\\"Is not A\\")
  return false
}",
  "defined-6.deep-1.defined-5.deep-0.rest-validator-6.explanations": Array [],
  "defined-6.deep-1.defined-5.explanations": Array [],
  "defined-6.deep-1.explanations": Array [],
  "defined-6.deep-1.rest-validator-5": "function validator(value) {
  validator.explanations = []
  if (value === \\"B\\") {
    return true
  }
  validator.explanations.push(\\"Is not B\\")
  return false
}",
  "defined-6.deep-1.rest-validator-5.explanations": Array [],
  "defined-6.explanations": Array [],
  "explanations": Array [],
  "rest-validator-4": "function (value) { return value === c; }",
  "rest-validator-4.explanations": Array [],
}
`;

exports[`compileObjectSchemaWithRest obj with constant 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-2'](value)) {
      validator.explanations.push(...validator['defined-2'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-2'](value[key])) {
        validator.explanations.push(...validator['rest-validator-2'].explanations)
        return false
      }
    }
    return true
  }",
  "__propsWithSchemasDict": Object {
    "f": true,
    "id": true,
    "inf": true,
    "minf": true,
    "nan": true,
    "str": true,
    "symb": true,
    "t": true,
    "zero": true,
  },
  "defined-2": "function validator(value) {
    if (!value) return false
    if (validator.__validValues.str[value.str] !== true) return false
    if (validator.__validValues.symb[value.symb] !== true) return false
    if (value.id !== 42) return false
    if (!Number.isNaN(value.nan)) return false
    if (value.inf !== Infinity) return false
    if (value.minf !== -Infinity) return false
    if (value.zero !== 0) return false
    if (value.f !== false) return false
    if (value.t !== true) return false
    return true
  }",
  "defined-2.__validValues": Object {
    "str": Object {
      "Andrew": true,
    },
    "symb": Object {
      Symbol(test): true,
    },
  },
  "defined-2.explanations": Array [],
  "explanations": Array [],
  "rest-validator-2": "function validator(value) {
  return typeof value === 'string'
}",
  "rest-validator-2.explanations": Array [],
}
`;

exports[`compileObjectSchemaWithRest obj with func 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-1'](value)) {
      validator.explanations.push(...validator['defined-1'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-1'](value[key])) {
        validator.explanations.push(...validator['rest-validator-1'].explanations)
        return false
      }
    }
    return true
  }",
  "__propsWithSchemasDict": Object {
    "id": true,
  },
  "defined-1": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (value.id !== 42) {
      validator.explanations.push(value.id)
      return false
    }
    return true
  }",
  "defined-1.explanations": Array [],
  "explanations": Array [
    "Is not A",
  ],
  "rest-validator-1": "function validator(value) {
  validator.explanations = []
  if (value === \\"A\\") {
    return true
  }
  validator.explanations.push(\\"Is not A\\")
  return false
}",
  "rest-validator-1.explanations": Array [
    "Is not A",
  ],
}
`;

exports[`compileObjectSchemaWithRest obj with variants 2`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-3'](value)) {
      validator.explanations.push(...validator['defined-3'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-3'](value[key])) {
        validator.explanations.push(...validator['rest-validator-3'].explanations)
        return false
      }
    }
    return true
  }",
  "__propsWithSchemasDict": Object {
    "gender": true,
    "grade": true,
  },
  "defined-3": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['gender-0'](value.gender)) {
      validator.explanations.push(...validator['gender-0'].explanations)
      return false
    }
    if (!validator['grade-0'](value.grade)) {
      validator.explanations.push(...validator['grade-0'].explanations)
      return false
    }
    return true
  }",
  "defined-3.explanations": Array [],
  "defined-3.gender-0": "function validator(value) {
    if (validator.__validValuesDict[value] === true) return true
    return false
  }",
  "defined-3.gender-0.__validValuesDict": Object {
    "female": true,
    "male": true,
  },
  "defined-3.gender-0.explanations": Array [],
  "defined-3.grade-0": "function validator(value) {
    if (value === 1) return true
    if (value === 2) return true
    if (value === 3) return true
    if (value === 4) return true
    if (value === 5) return true
    if (value === 6) return true
    if (value === 7) return true
    if (value === 8) return true
    if (value === 9) return true
    if (value === 10) return true
    if (value === 11) return true
    return false
  }",
  "defined-3.grade-0.explanations": Array [],
  "explanations": Array [
    "Is not A",
    "Is not B",
  ],
  "rest-validator-3": "function validator(value) {
    validator.explanations = []
    if (value === \\"A\\") return true;
    if (value === \\"B\\") return true;
    validator.explanations.push(\\"Is not A\\")
    validator.explanations.push(\\"Is not B\\")
    return false
  }",
  "rest-validator-3.explanations": Array [
    "Is not A",
    "Is not B",
  ],
}
`;

exports[`compileObjectSchemaWithRest obj: variant explanations 4`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['defined-7'](value)) {
      validator.explanations.push(...validator['defined-7'].explanations)
      return false
    }
    const keys = Object.keys(value)
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i]
      if (validator.__propsWithSchemasDict[key] === true) continue
      if (!validator['rest-validator-7'](value[key])) {
        validator.explanations.push(...validator['rest-validator-7'].explanations)
        return false
      }
    }
    return true
  }",
  "__propsWithSchemasDict": Object {
    "var": true,
  },
  "defined-7": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (!validator['var-0'](value.var)) {
      validator.explanations.push(...validator['var-0'].explanations)
      return false
    }
    return true
  }",
  "defined-7.explanations": Array [],
  "defined-7.var-0": "function validator(value) {
    validator.explanations = []
    if (value === \\"A\\") return true;
    if (value === \\"B\\") return true;
    validator.explanations.push(\\"Is not A\\")
    validator.explanations.push(\\"Is not B\\")
    return false
  }",
  "defined-7.var-0.explanations": Array [],
  "explanations": Array [
    "Is not C",
  ],
  "rest-validator-7": "function validator(value) {
  validator.explanations = []
  if (value === \\"C\\") {
    return true
  }
  validator.explanations.push(\\"Is not C\\")
  return false
}",
  "rest-validator-7.explanations": Array [
    "Is not C",
  ],
}
`;
